src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp

```diff
@@ -78,98 +83,38 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-        { ExecuteCommandlineKey, ShortcutAction::ExecuteCommandline },
-        { FindKey, ShortcutAction::Find },
-        { MoveFocusKey, ShortcutAction::MoveFocus },
-        { NewTabKey, ShortcutAction::NewTab },
-        { NextTabKey, ShortcutAction::NextTab },
-        { OpenNewTabDropdownKey, ShortcutAction::OpenNewTabDropdown },
-        { OpenSettingsKey, ShortcutAction::OpenSettings },
-        { OpenTabColorPickerKey, ShortcutAction::OpenTabColorPicker },
-        { PasteTextKey, ShortcutAction::PasteText },
-        { PrevTabKey, ShortcutAction::PrevTab },
-        { RenameTabKey, ShortcutAction::RenameTab },
-        { OpenTabRenamerKey, ShortcutAction::OpenTabRenamer },
-        { ResetFontSizeKey, ShortcutAction::ResetFontSize },
-        { ResizePaneKey, ShortcutAction::ResizePane },
-        { ScrolldownKey, ShortcutAction::ScrollDown },
-        { ScrolldownpageKey, ShortcutAction::ScrollDownPage },
-        { ScrollupKey, ShortcutAction::ScrollUp },
-        { ScrolluppageKey, ShortcutAction::ScrollUpPage },
-        { ScrollToTopKey, ShortcutAction::ScrollToTop },
-        { ScrollToBottomKey, ShortcutAction::ScrollToBottom },
-        { SendInputKey, ShortcutAction::SendInput },
-        { SetColorSchemeKey, ShortcutAction::SetColorScheme },
-        { SetTabColorKey, ShortcutAction::SetTabColor },
-        { SplitPaneKey, ShortcutAction::SplitPane },
-        { SwitchToTabKey, ShortcutAction::SwitchToTab },
-        { TabSearchKey, ShortcutAction::TabSearch },
-        { ToggleAlwaysOnTopKey, ShortcutAction::ToggleAlwaysOnTop },
-        { ToggleCommandPaletteKey, ShortcutAction::ToggleCommandPalette },
-        { ToggleFocusModeKey, ShortcutAction::ToggleFocusMode },
-        { ToggleFullscreenKey, ShortcutAction::ToggleFullscreen },
-        { TogglePaneZoomKey, ShortcutAction::TogglePaneZoom },
-        { LegacyToggleRetroEffectKey, ShortcutAction::ToggleShaderEffects },
-        { ToggleShaderEffectsKey, ShortcutAction::ToggleShaderEffects },
-        { MoveTabKey, ShortcutAction::MoveTab },
-        { BreakIntoDebuggerKey, ShortcutAction::BreakIntoDebugger },
-        { UnboundKey, ShortcutAction::Invalid },
-        { FindMatchKey, ShortcutAction::FindMatch },
-        { TogglePaneReadOnlyKey, ShortcutAction::TogglePaneReadOnly },
-        { NewWindowKey, ShortcutAction::NewWindow },
-        { IdentifyWindowKey, ShortcutAction::IdentifyWindow },
-        { IdentifyWindowsKey, ShortcutAction::IdentifyWindows },
-        { RenameWindowKey, ShortcutAction::RenameWindow },
-        { OpenWindowRenamerKey, ShortcutAction::OpenWindowRenamer },
-        { GlobalSummonKey, ShortcutAction::GlobalSummon },
-        { QuakeModeKey, ShortcutAction::QuakeMode },
+
+#define ON_ALL_ACTIONS(action) KEY_TO_ACTION_PAIR(action)
+        ALL_SHORTCUT_ACTIONS
+#undef ON_ALL_ACTIONS
+    };
+
+    static const std::map<ShortcutAction, std::string_view, std::less<>> ActionToStringMap{
```

Comment:
We have got to stop it with the globally initialized maps. Sorry. This looks like a case for til::static_map or til::presorted_static_map. I'd rather do that than have yet another static initializer :smile:

Reply:
Should I do what we did with `GeneratedActionNames` in `ActionAndArgs::GenerateName`? Define these in the function they're used as static lambda functions. Or does that not make this any better?

Reply:
@DHowett Can you explain the problem with static initializers for the uninitiated? 🙂

Reply:
Sure!

Initializing a static _non-constexpr_ value at file scope generates a block of code that must be executed when the library is loaded, every time it is loaded. For something like the settings model, that cost is paid once on startup, but it is still a cost that _must_ be paid before any of the actual settings model code is run.

constexpr is a different story: those are evaluated at compile time and stored permanently in the binary.

For a really, really good example of the cost of a static initializer (or something very similar) and the payback you get from replacing it with a fully-constexpr data store, look at commit 16e1e29a12ac3f3f2f41d0a24be41dedab1f5605.

Reply:
This seems fairly solid - it even offers perfect hash functions: https://github.com/serge-sans-paille/frozen

Reply:
Leonard and I couldn't get this to work, unfortunately. We kept running into `til::presorted_static_map` saying that it couldn't find the `static_map` in the ctor or something like that. Sorry!